New package: py-x21

qwertos opened the pull request:
@adamjstewart I know there is going to be a couple points here where you're going to want me to do some fixes. I'm submitting this early as a draft to ask for some guidance.

### 1. Description

Currently: 

```python
    """FIXME: I think this is for unpacking this authors
    obfuscaded python libraries."""
```

The package itself doesn't provide a description, so I'm basing this on how this gets called in some dependent packages (e.g. [colorio](https://pypi.org/project/colorio/)). Is this speculation good enough? And do you concur with my interpretation of it's use? (Regardless, it's going to be word smithed a bit to remove the "FIXME:" and "I think")

### 2. The checksum/URL situation

Each version has multiple different wheels depending on the platform being used and version of python. Currently I'm doing:

```python
    # NOTE: Due to the this package having different packages for each
    # version of python and platform type, checksums have been proveded
    # here. Editing the version line below will likely be required to
    # build on your system.
    checksums = {
        '0.2.6': {
            'linux': {
                '3.7': '8b35248d0b049dd09985d1a45c6fa36dd39db2c9d805a96028ec3bf9dc80e0dd',
                '3.8': '64275052bcda784395bc613f750b8b5a6b1ddbfa4e7a590cb8e209543f0ca0c4',
                '3.9': 'e20b29650fcbf0be116ac93511033bf10debc76261b7350e018ff91b92ff950d',
                '3.10': '7c5c58ff6dc81caac6815578f78cf545e719beb0bf4017f77120d38025d2bc7d'},
            'darwin': {}}}

    version('0.2.6', sha256=checksums['0.2.6']['linux']['3.8'], expand=False)
```

Resolving this issue by adding something along the lines of the following, would make the most sense to me; but I don't think it's currently possible. (Correct me if I'm wrong.)

```python
    def sha256_for_version(self, version):
        return checksums[version][self.spec.platform][self.spec['python'].version.up_to(2)]
```

How do you recommend dealing with this issue?

vivian-hafener committed:
Py-x21 now works, needs dependencies

Conflicts:
	var/spack/repos/rit-rc/packages/py-x21/package.py

vivian-hafener committed:
Added dependencies to py-x21

vivian-hafener committed:
Making flake style check happy

qwertos committed:
[py-x21] flake8

qwertos committed:
[py-x21]

- added homepage
- added placeholder description
- added comment about checksums

qwertos committed:
[py-x21] added darwin support and fixed issue with python 3.7 wheel name

qwertos committed:
[py-x21] adding checksum hash

adamjstewart commented on var/spack/repos/builtin/packages/py-x21/package.py:
```diff
@@ -0,0 +1,49 @@
+    depends_on('py-setuptools', type=('build', 'run'))
+    depends_on('py-tomli', type=('build', 'run'))
+    depends_on('py-tomli-w', type=('build', 'run'))
+    depends_on('py-pynacl', type=('build', 'run'))
```
This line is duplicated

adamjstewart commented on var/spack/repos/builtin/packages/py-x21/package.py:
```diff
@@ -0,0 +1,49 @@
+
+class PyX21(PythonPackage):
+    """FIXME: I think this is for unpacking this authors
+    obfuscaded python libraries."""
```
No idea what to do about the description, I think this is fine

adamjstewart commented on var/spack/repos/builtin/packages/py-x21/package.py:
```diff
@@ -0,0 +1,49 @@
+                '3.10': '7c5c58ff6dc81caac6815578f78cf545e719beb0bf4017f77120d38025d2bc7d'},
+            'darwin': {}}}
+
+    version('0.2.6', sha256=checksums['0.2.6']['linux']['3.8'], expand=False)
```
These kinds of packages are a real pain to deal with. In the past my strategy has been what you see in packages like `py-azureml-dataprep-native`, `py-azureml-dataprep-rslex`, and `py-itk`. I would be fine with a solution like that, or a new solution you come up with. I agree that we really need a `when` parameter for `version` to support this correctly but I remember that being more difficult to add than expected.

adamjstewart reviewed (commented):
Hope these comments help

qwertos committed:
[py-x21] removed duplicate py-pynacl

qwertos committed:
[py-x21]

- updated description
- updated version listing to have a different version for each version
  of python. Also, versions dependent on sys.platform
- updated url_for_version to not require post concretized information so
  that spack checksum works

qwertos committed:
[py-x21] isort

adamjstewart reviewed (approved):
LGTM, is this still a WIP or can it be merged?
